File: repo-file-sync.ts
import { ConfigParser } from './libs/config-parser.js';
import { GitHubFetcher } from './libs/github-fetcher.js';
import type { Config, RepoSource } from './types.js';

export class RepoFileSync {
  private configParser = new ConfigParser();
  private fetcher = new GitHubFetcher();

  async sync(configPath: string): Promise<void> {
    console.log(`Reading config from ${configPath}...`);
    const config = await this.configParser.parseConfig(configPath);
    
    console.log(`Found ${config.sources.length} source(s) to sync`);
    
    for (const source of config.sources) {
      await this.syncSource(source);
    }
    
    console.log('Sync completed successfully!');
  }

  private async syncSource(source: RepoSource): Promise<void> {
    console.log(`\nSyncing from ${source.repo} (ref: ${source.ref})`);
    
    if (source.vars && source.vars.length > 0) {
      console.log(`  Template variables: ${source.vars.map(v => `${v.key} → ${v.value}`).join(', ')}`);
    }
    
    for (const file of source.files) {
      try {
        console.log(`  Downloading ${file}...`);
        await this.fetcher.downloadFile(source.repo, source.ref, file, undefined, source.vars);
      } catch (error) {
        console.error(`  Failed to download ${file}: ${error}`);
        throw error;
      }
    }
  }
}

File: repo_file_sync.py
from libs.config_parser import ConfigParser
from libs.github_fetcher import GitHubFetcher
from types import Config, RepoSource


class RepoFileSync:
    def __init__(self):
        self.config_parser = ConfigParser()
        self.fetcher = GitHubFetcher()

    def sync(self, config_path: str) -> None:
        print(f"Reading config from {config_path}...")
        config = self.config_parser.parse_config(config_path)
        
        print(f"Found {len(config['sources'])} source(s) to sync")
        
        for source in config['sources']:
            self._sync_source(source)
        
        print('Sync completed successfully!')

    def _sync_source(self, source: RepoSource) -> None:
        print(f"\nSyncing from {source['repo']} (ref: {source['ref']})")
        
        for file in source['files']:
            try:
                print(f"  Downloading {file}...")
                self.fetcher.download_file(source['repo'], source['ref'], file)
            except Exception as error:
                print(f"  Failed to download {file}: {error}")
                raise error